### ISSUE TITLE
Improve reverse() algorithm performance

### ISSUE THREAD

**Issue by kylelutz** [kylelutz/compute#25]
Take a tile-based approach, reverse blocks in local memory and then write to global memory.

**Comment by haahh** [comment]
I've implemented local memory solution (mentioned above) - https://github.com/haahh/compute/commit/c2e94838c6b478c60e60ac0de6911102cd5622e8. It is just first version and ofc needs refactor etc.

More important is that the performance is virtually the same (sometimes lower, sometimes higher) compared to current solution (only global memory) on my device (AMD Radeon HD7770). Also global memory solution is more stable in terms of performance. When buffer is small (<10000) difference is noticeable to the disadvantage of local memory solution.

I tested it with CodeXL GPU Performance counters (running perf_reverse), 10 tests for each version of kernel.

I think it's because with current solution reading and writing from global memory is already coalesced/contiguous/adjacent. 

```c++
   decl<cl_uint>("i") << " = get_global_id(0);\n" <<
   decl<cl_uint>("j") << " = size - get_global_id(0) - 1;\n" <<
   decl<value_type>("tmp") << "=" << first[var<cl_uint>("i")] << ";\n" <<
   first[var<cl_uint>("i")] << "=" << first[var<cl_uint>("j")] << ";\n" <<
   first[var<cl_uint>("j")] << "= tmp;\n";
```

I mean this part ` decl<value_type>("tmp") << "=" << first[var<cl_uint>("i")] << ";\n"` is obviously coalesced, textbook example. And this part ` first[var<cl_uint>("i")] << "=" << first[var<cl_uint>("j")] << ";\n"` is read/write in reversed manner. Device OpenCL compiler and/or device drivers (I dont know details) know what to do so it would be coalesced.

Probably with older devices (NVIDIA GPU with CC 1.0/1.1) using local memory would be some difference. I will test it later.
